pydantic/generics.py

```diff
@@ -0,0 +1,85 @@
+from typing import Any, ClassVar, Dict, Generic, Tuple, Type, TypeVar, Union, get_type_hints
+
+from pydantic import BaseModel, create_model
+from pydantic.class_validators import gather_validators
+
+_generic_types_cache: Dict[Tuple[Type[Any], Union[Any, Tuple[Any, ...]]], Type[BaseModel]] = {}
+GenericModelT = TypeVar('GenericModelT', bound='GenericModel')
+
+
+class GenericModel(BaseModel):
+    __slots__ = ()
+    __concrete__: ClassVar[bool] = False
+
+    def __new__(cls, *args: Any, **kwargs: Any) -> Any:
+        if cls.__concrete__:
+            return super().__new__(cls)
+        else:
+            raise TypeError(f'Type {cls.__name__} cannot be used without generic parameters, e.g. {cls.__name__}[T]')
+
+    def __class_getitem__(  # type: ignore
+        cls: Type[GenericModelT], params: Union[Type[Any], Tuple[Type[Any], ...]]
+    ) -> Type[BaseModel]:
+        cached = _generic_types_cache.get((cls, params))
+        if cached is not None:
+            return cached
+        if cls.__concrete__:
+            raise TypeError('Cannot parameterize a concrete instantiation of a generic model')
+        if not isinstance(params, tuple):
+            params = (params,)
+        if any(isinstance(param, TypeVar) for param in params):  # type: ignore
+            raise TypeError(f'Type parameters should be placed on typing.Generic, not GenericModel')
+        if Generic not in cls.__bases__:
```

Comment:
It's important to put this check for inheriting from Generic *after* the check for TypeVar parameterization, or the error messages end up being more confusing if you try to put type vars directly on the GenericModel (which I would expect to be a relatively common mistake).